### PR TITLE
[base 1.11.x beta] Sync: Display a deprecation warning for users of sync v1

### PR DESCRIPTION
**Pull request by petemill**
Resolves https://github.com/brave/brave-browser/issues/10231

### A
- Enable `brave://flags` flag for brave sync
- Restart
- Create a sync chain at `brave://sync`
- Restart
- **Observe** infobar is shown:
image

### B
- Perform Plan A
- Click link
- **Observe** infobar is removed from source tab
- Restart
- **Observe** Infobar is not shown

### C
- Perform Plan A
- Click (x)
- Restart
- **Observe** Infobar is not shown

## Submitter Checklist:

- [x] Submitted a [ticket](https://github.com/brave/brave-browser/issues) for my issue if one did not already exist.
- [x] Used Github [auto-closing keywords](https://help.github.com/articles/closing-issues-via-commit-messages/) in the commit message.
- [ ] Added/updated tests for this change (for new code or code which already has tests).
- Verified that these changes build without errors on
  - [ ] Android
  - [ ] iOS
  - [ ] Linux
  - [ ] macOS
  - [ ] Windows
- Verified that these changes pass automated tests (unit, browser, security tests) on
  - [ ] iOS
  - [ ] Linux
  - [ ] macOS
  - [ ] Windows
- [x] Verified that all lint errors/warnings are resolved (`npm run lint`)
- [ ] Ran `git rebase master` (if needed).
- [ ] Ran `git rebase -i` to squash commits (if needed).
- [ ] Tagged reviewers and labelled the pull request as needed.
- [ ] Request a security/privacy review as needed.
- [ ] Add appropriate QA labels (QA/Yes or QA/No) to include the closed issue in milestone
- [ ] Public documentation has been updated as necessary. For instance:
  - [ ] https://github.com/brave/brave-browser/wiki/Deviations-from-Chromium-(features-we-disable-or-remove)
  - [ ] https://github.com/brave/brave-browser/wiki/Proxy-redirected-URLs
  - [ ] https://github.com/brave/brave-browser/wiki/Fingerprinting-Protections
  - [ ] https://github.com/brave/brave-browser/wiki/Brave%E2%80%99s-Use-of-Referral-Codes
  - [ ] https://github.com/brave/brave-browser/wiki/Custom-Headers
  - [ ] https://github.com/brave/brave-browser/wiki/Web-compatibility-issues-with-tracking-protection
  - [ ] https://github.com/brave/brave-browser/wiki/QA-Guide

## Reviewer Checklist:

- [ ] New files have MPL-2.0 license header.
- [ ] Request a security/privacy review as needed.
- [ ] Adequate test coverage exists to prevent regressions 
- [ ] Verify test plan is specified in PR before merging to source

## After-merge Checklist:

- [ ] The associated issue milestone is set to the smallest version that the
  changes has landed on.
- [ ] All relevant documentation has been updated.

### CODE REVIEW

File: app/brave_generated_resources.grd [hunk 1]
```diff
@@ -479,6 +479,10 @@ By installing this extension, you are agreeing to the Google Widevine Terms of U
       <message name="IDS_SETTINGS_HELP_TIPS_SHOW_BRAVE_WAYBACK_MACHINE_PROMPT" desc="The label describing the 'wayback machine' toggle option">
         Show Wayback Machine prompt on 404 pages
       </message>
+      <!-- Sync Infobar -->
+      <message name="IDS_BRAVE_SYNC_V1_DEPRECATED_INFOBAR_MESSAGE" desc="Message text shown to users who used a previous version of Brave Sync and need to learn about a new version coming soon.">
+        A new Brave Sync is coming. Get ready for the upgrade.
```
Comment (darkdh) [inline app/brave_generated_resources.grd:484]: I don't see this message telling user that sync v1 will soon be no longer available. Or will user see that in https://brave.com/sync-v2-is-coming?

Comment (petemill) [reply app/brave_generated_resources.grd:484]: Yeah it should be there I suppose. But I'll bring up the question as to whether we should show this to users who are currently _syncing_, i.e. have manually turned the sync v1 flag on.

Comment (petemill) [reply app/brave_generated_resources.grd]: Made the following changes based on feedback:

- Updated the message according to @jsecretan's suggestion:
> A new Brave Sync is coming and will affect your setup. Get ready for the upgrade.

- This infobar only shows to users who have re-enabled sync v1 via the flag.

File: browser/infobars/sync_v1_deprecation_infobar_delegate.cc [hunk 2]
```diff
@@ -0,0 +1,92 @@
...
+  if (!is_flag_enabled) {
+    return;
+  }
+  if (!prefs->GetBoolean(brave_sync::prefs::kSyncEnabled) ||
+      prefs->GetBoolean(
+          brave_sync::prefs::kSyncDeprecationWarningNoticeDismissed)) {
+    return;
+  }
+
+  // Only show the bar once, ever
```
Comment (simonhong) [inline browser/infobars/sync_v1_deprecation_infobar_delegate.cc:44]: If user uses multiple profile, maybe that user could see this infobar multiple times.

Comment (petemill) [reply browser/infobars/sync_v1_deprecation_infobar_delegate.cc:44]: I'm ok with that, the other profiles must have sync enabled too. We can't assume that the other profiles are not other people perhaps.

File: browser/infobars/sync_v1_deprecation_infobar_delegate.cc [hunk 3]
```diff
@@ -0,0 +1,92 @@
+/* Copyright (c) 2019 The Brave Authors. All rights reserved.
```
Comment (simonhong) [inline browser/infobars/sync_v1_deprecation_infobar_delegate.cc:1]: nit: 2020

File: browser/infobars/sync_v1_deprecation_infobar_delegate.h [hunk 4]
```diff
@@ -0,0 +1,38 @@
+/* Copyright (c) 2019 The Brave Authors. All rights reserved.
```
Comment (simonhong) [inline browser/infobars/sync_v1_deprecation_infobar_delegate.h:1]: nit: 2020

**Comment by kjozwiak** [comment]
Looks like `test-install` failed on `macOS` as per https://ci.brave.com/job/pr-brave-browser-sync-v1-deprecation-notice/4/execution/node/647/log/. However, this a known intermittent failure and doesn't block uplift.

**Pull request merged**